espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/MobileBackdoor.java

```
@@ -0,0 +1,110 @@
+        }
+
+        return invocationResult.toString();
+    }
+
+    private Object invokeOnActivity(Activity activity, List<InvocationOperation> ops) throws AppiumException {
+        Object invocationResult = null;
+        Object invokeOn = activity;
+        for (InvocationOperation op : ops) {
+            try {
+                invocationResult = op.apply(invokeOn);
+                invokeOn = invocationResult;
+            } catch (Exception e) {
+                throw new AppiumException(e);
+            }
+        }
+        return invocationResult;
+    }
+
+    private Object safeInvokeOnApplication(Activity activity, List<InvocationOperation> ops) {
+        Object invocationResult = null;
+        Object invokeOn = activity.getApplication();
+        for (InvocationOperation op : ops) {
+            try {
+                invocationResult = op.apply(invokeOn);
+                invokeOn = invocationResult;
+            } catch (Exception e) {
+                e.printStackTrace();
+            }
+        }
+        return invocationResult;
+    }
+
+    private List<InvocationOperation> getBackdoorOperations(MobileBackdoorParams params) throws InvalidArgumentException {
+        List<InvocationOperation> ops = new ArrayList<>();
+        List<MobileBackdoorMethod> mobileBackdoorMethods = params.getOpts();
+
+        for (MobileBackdoorMethod mobileBackdoorMethod : mobileBackdoorMethods) {
+            String methodName = mobileBackdoorMethod.getName();
+            if (methodName == null) {
+                throw new InvalidArgumentException("'name' is a required parameter for backdoor method to be invoked.");
+            }
+            ops.add(new InvocationOperation(methodName, mobileBackdoorMethod.getArgs()));
+        }
+        return ops;
+    }
+
+
+    //    https://androidreclib.wordpress.com/2014/11/22/getting-the-current-activity/
+    private Activity getCurrentActivity() throws AppiumException {
```

This can be a separate static helper method. I think it makes sence to move it into a separate utility class and make public, since it might be useful in other classes. Also it should throw an exception instead of returning null if the current activity is not found

---

done